nucypher/blockchain/eth/actors.py

```diff
@@ -275,82 +341,124 @@ class Miner(NucypherTokenActor):
             self.stake()
 
     #
     # Staking
     #
+
     @only_me
     def stake(self, confirm_now: bool = True) -> None:
-        """High-level staking looping call initialization"""
-        # TODO #841: Check if there is an active stake in the current period: Resume staking daemon
-
+        """
+        High-level staking looping call initialization, this function aims
+        to be safely called at any time - For example, it is okay to call
+        this function multiple time within the same period.
+        """
         # Get the last stake end period of all stakes
-        terminal_period = max(stake.end_period for stake in self.stakes.values())
+        terminal_period = max(stake.end_period for stake in self.stakes)
 
         if confirm_now:
             self.confirm_activity()
 
         # record start time and periods
         self.__start_time = maya.now()
         self.__uptime_period = self.miner_agent.get_current_period()
-        self.__terminal_period = self.__uptime_period + terminal_period
+        self.__terminal_period = terminal_period
         self.__current_period = self.__uptime_period
         self.start_staking_loop()
 
+    @property
+    def last_active_period(self) -> int:
+        period = self.miner_agent.get_last_active_period(address=self.checksum_public_address)
+        return period
+
     @only_me
     def _confirm_period(self):
 
-        period = self.miner_agent.get_current_period()
+        onchain_period = self.miner_agent.get_current_period()  # < -- Read from contract
         self.log.info("Checking for new period. Current period is {}".format(self.__current_period))
 
-        if self.__current_period != period:
+        # Check if the period has changed on-chain
+        if self.__current_period != onchain_period:
 
-            # check for stake expiration
+            # Let's see how much time has passed
+            # TODO: Follow-up actions for downtime
+            missed_periods = onchain_period - self.last_active_period
+            if missed_periods:
+                self.log.warn(f"MISSED CONFIRMATION - {missed_periods} missed staking confirmations detected!")
+                self.__read_stakes()  # Invalidate the stake cache
+
+            # Check for stake expiration and exit
             stake_expired = self.__current_period >= self.__terminal_period
             if stake_expired:
-                self.log.info('Stake duration expired')
+                self.log.info('STOPPED STAKING - Final stake ended.')
                 return True
 
+            # Write to Blockchain
             self.confirm_activity()
-            self.__current_period = period
+
+            # Update local period cache
+            self.__current_period = onchain_period
             self.log.info("Confirmed activity for period {}".format(self.__current_period))
 
-    @only_me
+    def heartbeat(self):
+        """Used with LoopingCall"""
+        try:
```

Comment:
If we reraise anyway, why not simply `self._confirm_period()`?

Reply:
This was more or less me stubbing a future approach that aggregates multiple looping calls and manages them in sequence.  Unsure if this is a good design direction, thoughts?.  I can remove it from this PR for now since its out of scope.

Reply:
It's ok to leave here if you leave a comment, what it is for